lib/api/apiUtils/bucket/models/ReplicationConfiguration.js

```diff
@@ -0,0 +1,307 @@
+        this.replicationConfiguration = {
+            role: null,
+            destination: null, // The bucket to replicate data to.
+            rules: [
+                {
+                    storageClass: null,
+                    prefix: null,
+                    status: null,
+                    id: null,
+                },
+            ],
+        };
+    }
+
+    /**
+     * Build the rule object from the parsed XML of the given rule
+     * @param {object} rule - The rule object from this._parsedXML
+     * @return {object} - The rule object to push into the `Rules` array
+     */
+    _getRuleObject(rule) {
+        const obj = {
+            prefix: rule.Prefix[0],
+            status: rule.Status[0],
+        };
+        // StorageClass is an optional property.
+        if (rule.Destination[0].StorageClass) {
+            obj.storageClass = rule.Destination[0].StorageClass[0];
+        }
+        // ID is an optional property.
+        if (rule.ID) {
+            obj.id = rule.ID[0];
+        }
+        return obj;
+    }
+
+    /**
+     * Check that the `Role` property of the configuration is valid
+     * @return {undefined}
+     */
+    _parseRole() {
+        const Role = this._parsedXML.ReplicationConfiguration.Role;
+        if (!Role) {
+            return errors.MalformedXML;
+        }
+        // TODO: Update to validate role priveleges after implemented in Vault.
+        // Role should be an IAM user name.
+        const arr = Role[0].split(':');
+        const isValidARN = arr.length === 7 ||
+            (arr.length === 6 && arr[5].split('/').length === 2);
+        if (!isValidARN) {
+            return errors.InvalidArgument.customizeDescription(
+                'Invalid Role specified in replication configuration');
+        }
+        this.replicationConfiguration.role = Role[0];
+        return undefined;
+    }
+
+    /**
+     * Check that the `Rules` property array is valid
+     * @return {undefined}
+     */
+    _parseRules() {
+        // Note that the XML uses 'Rule' while the config object uses 'Rules'.
+        const { Rule } = this._parsedXML.ReplicationConfiguration;
+        if (!Rule || Rule.length < 1) {
+            return errors.MalformedXML;
+        }
+        if (Rule.length > 1000) {
+            return errors.InvalidRequest.customizeDescription(
+                'Number of defined replication rules cannot exceed 1000');
+        }
+        const err = this._parseEachRule(Rule);
+        if (err) {
+            return err;
+        }
+        this.replicationConfiguration.rules = this._rules;
+        return undefined;
+    }
+
+    /**
+     * Check that each rule in the `Rules` property array is valid
+     * @param {array} rules - The rule array from this._parsedXML
+     * @return {undefined}
+     */
+    _parseEachRule(rules) {
+        const rulesArr = [];
+        for (let i = 0; i < rules.length; i++) {
+            const err =
```

Comment:
This is not an error more like a `check`. Let's reserve `err` variable name for Error instance objects.